### PR TITLE
feat: Add linear CE loss fusion for DPO

### PR DESCRIPTION
**Pull request by pengdurice**
# What does this PR do ?

**Suport Linear CE Loss Fusion for DPO**
On top of https://github.com/NVIDIA-NeMo/RL/pull/2036 where Linear CE loss fusion support is added for SFT. This PR adds the support to DPO loss.
# Optimizations

## Chunked Linear Cross-Entropy Fusion Loss

During standard DPO training the model materializes a full logit tensor of shape `[batch_size, seq_length, vocab_size]` (up to parallelism) for both the policy forward-backward pass and the reference model logprob computation. This can cause out-of-memory (OOM) errors for long sequences or large vocabularies. The **chunked linear cross-entropy fusion loss** avoids this by computing log probabilities directly from the hidden states: it chunks the sequence dimension, projects each chunk to logits on the fly, gathers per-token log probabilities, and discards the logits before moving to the next chunk.

**Benefits:**

- Extends the maximum trainable sequence length significantly by eliminating the large logit tensor from GPU memory.
- Applies to both the training forward-backward pass and the reference model logprob computation.
- Produces numerically equivalent loss values to the standard path.
# Issues
NA

# Tests
1. Unit tests passed
2. Local tests (running the sh file) passed
3. Extended context window from <30K to 40K. Even larger context window triggers OOM in another place which will be fixed in another PR.
4. A side by side loss curve comparison between baseline and experiment(linear ce loss fusion enabled)
image

# Usage

```python
# add the following to your dpo.yaml file.
megatron_cfg:
    enabled: true
    use_linear_ce_fusion_loss: true
    linear_ce_fusion_chunk_size: 256 # or other numbers 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [X] Make sure you read and followed [Contributor guidelines](/NVIDIA-NeMo/RL/blob/main/CONTRIBUTING.md)
- [X] Did you write any new necessary tests?
- [X] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA-NeMo/RL/blob/main/docs/testing.md) for how to run tests
- [x] Did you add or update any necessary documentation? Visit our [Document Development Guide](/NVIDIA-NeMo/RL/blob/main/docs/documentation.md) for how to write, build and test the docs.

# Additional Information
* ...

## Release Notes

* **Documentation**
  * Added DPO training optimization guide section
  * Updated SFT training documentation to reference optimization

* **New Features**
  * Added example DPO configuration for Qwen2.5-Math-7B model
  * Enabled chunked linear cross-entropy loss support for DPO training

* **Tests**
  * Added nightly test suite for DPO with chunked linear cross-entropy loss
  * Added unit test validating chunked linear cross-entropy loss behavior

### CODE REVIEW

File: nemo_rl/models/megatron/train.py [hunk 1]
```diff
@@ -427,10 +429,13 @@ def __call__(
         original_seq_length = unpacked_input_ids.shape[1]
 
         def processor_fn_inner(output_tensor):
-            tp_grp = get_tensor_model_parallel_group()
-            tp_rank = get_tensor_model_parallel_rank()
-            logprob_chunk_size = self.cfg.get("logprob_chunk_size", None)
-            if self.cfg["sequence_packing"]["enabled"]:
+            if self.use_linear_ce_fusion:
+                token_logprobs = output_tensor.to(torch.float32)
+                token_logprobs = token_logprobs[:, : original_seq_length - 1]
```
Comment (yuki-97) [inline nemo_rl/models/megatron/train.py:434]: am I understand correctly that `use_linear_ce_fusion` here works for both sequence_packing is True or False?

Comment (pengdurice) [reply nemo_rl/models/megatron/train.py:434]: Hi @yuki-97 , thank you for your question!  Nope, if I understand correctly, currently, DPO and sequence packing are mutually exclusive: https://github.com/NVIDIA-NeMo/RL/blob/b0da493855a39ff781547339bf64eb01160f9582/nemo_rl/algorithms/dpo.py#L133 So here, we first check if loss fusion is enabled, if so, go to that path, elif and else will go to original path. LMK if that makes sense;-)

Comment (yuki-97) [reply nemo_rl/models/megatron/train.py:434]: oh I see, thanks for pointing this.

wdyt about adding an assert after that? so that if we support sequence packing in dpo, people can know we can't use sequence packing and loss fusion together.
```python
if master_config["policy"]["sequence_packing"]["enabled"]:
    assert xxx  # assert not using loss fusion
```

Comment (pengdurice) [reply nemo_rl/models/megatron/train.py:434]: sounds a great idea! Added this guardrail and we can remove it once DPO + sequence packing + linear fusion loss all 3 are compatible.

File: docs/guides/dpo.md [hunk 2]
```diff
@@ -195,6 +195,36 @@ The DPO implementation in NeMo RL supports several key parameters that can be ad
...
+policy:
+  megatron_cfg:
+    enabled: true
+    use_linear_ce_fusion_loss: true
+    linear_ce_fusion_chunk_size: 256  # tokens per chunk; smaller = less memory, larger = more throughput
+```
+
+**Notes:**
+
+- Context parallelism is not supported when linear CE fusion is enabled.
```
Comment (terrykong) [inline docs/guides/dpo.md:224]: is there an assert that checks for this?

Comment (pengdurice) [reply docs/guides/dpo.md:224]: Hi @terrykong , thank you for checking! https://github.com/NVIDIA-NeMo/RL/blob/main/nemo_rl/models/megatron/setup.py#L381 it was added in the previous PR.  https://github.com/NVIDIA-NeMo/RL/pull/2036 Thanks!

Comment (terrykong) [reply docs/guides/dpo.md:224]: disregard. @yuki-97 reminded me that it was already done in the SFT PR https://github.com/NVIDIA-NeMo/RL/blob/441005510d56a11067d44fb6314c3b3f193619b7/nemo_rl/models/megatron/setup.py#L377-L383
